pkg/apis/resource/types.go

```diff
@@ -37,90 +298,329 @@ type ResourceClaim struct {
 	// +optional
 	metav1.ObjectMeta
 
-	// Spec describes the desired attributes of a resource that then needs
-	// to be allocated. It can only be set once when creating the
-	// ResourceClaim.
+	// Spec describes what is being requested and how to configure it.
+	// The spec is immutable.
 	Spec ResourceClaimSpec
 
-	// Status describes whether the resource is available and with which
-	// attributes.
+	// Status describes whether the claim is ready to use and what has been allocated.
 	// +optional
 	Status ResourceClaimStatus
 }
 
-// ResourceClaimSpec defines how a resource is to be allocated.
+// ResourceClaimSpec defines what is being requested in a ResourceClaim and how to configure it.
 type ResourceClaimSpec struct {
-	// ResourceClassName references the driver and additional parameters
-	// via the name of a ResourceClass that was created as part of the
-	// driver deployment.
-	ResourceClassName string
+	// Devices defines how to request devices.
+	//
+	// +optional
+	Devices DeviceClaim
```

Comment:
Ok - don't hate me, but I have to ask this.

Now that we have put almost everything under `DeviceClaim`, maybe we should make that the top-level object, and get rid of `ResourceClaim`? This would allows separate RBAC for device claims and gizmos claims. The capacity side (`ResourceSlice`) would stay as-is, I am just talking about the claim side.

Reply:
@pohly and I came to the same conclusion yesterday, but realized it would introduce alot of churn to make this change at this point. I am not opposed to it (it definitely makes the Class/Claim more symmetrical). But we should do it as a follow-up PR once this one gets merged.

Reply:
My arguments for a generic ResourceClaim type still hold:
- There is code which doesn't need to know about "devices", only about claims.
- V1 core API only needs to be changed once.
- If we want to define constraints that cover both "devices" and future "gizmos", then we can add them if both are in the same claim. We can't do that as easily across claims for the same reason that we currently don't do cross-claim device constraints (would need to be configured in the pod spec).

> it definitely makes the Class/Claim more symmetrical

But that would be a false symmetry. A device request inside a claim matches with a DeviceClass, not the claim. DeviceClass is thus one level down in the layering compared to ResourceClaim.

Reply:
Let me clarify my statement above....

I'm not saying I'm 100% on board with changing ResourceClaim --> DeviceClaim. I actually don't feel too strongly either way, as there are valid arguments on both sides. 

What I meant to say was -- _**if**_ we decide we want to pursue this change, let's do it as a follow-up PR once this PR gets merged.

Reply:
Sure. Something to consider later. I agree there are pros and cons both ways and probably not enough insight into the future to say.